From 2b84a22a0d84e5d07cbaa0a406eec15fd5fb10ec Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Caol=C3=A1n=20McNamara?= <caolanm@redhat.com>
Date: Thu, 13 Aug 2015 10:58:06 +0100
Subject: [PATCH] convert pStatus to vector and use at to check offsets

(cherry picked from commit ea70088895ed45dc60abf18319acc1b4fa3018dd)

Change-Id: I5186f6a65bb9d5ed8a0d1ab1d71f7e2c13865411
Reviewed-on: https://gerrit.libreoffice.org/17695
Reviewed-by: David Tardon <dtardon@redhat.com>
Tested-by: David Tardon <dtardon@redhat.com>
DebianPatchName: pstatus-vector.diff
---
 sw/source/filter/ww8/ww8scan.cxx | 16 ++++++++--------
 sw/source/filter/ww8/ww8scan.hxx |  4 ++--
 2 files changed, 10 insertions(+), 10 deletions(-)

diff --git a/sw/source/filter/ww8/ww8scan.cxx b/sw/source/filter/ww8/ww8scan.cxx
index 06993b4..3d862c8 100644
--- a/sw/source/filter/ww8/ww8scan.cxx
+++ b/sw/source/filter/ww8/ww8scan.cxx
@@ -3972,7 +3972,7 @@ void WW8ReadSTTBF(bool bVer8, SvStream& rStrm, sal_uInt32 nStart, sal_Int32 nLen
 }
 
 WW8PLCFx_Book::WW8PLCFx_Book(SvStream* pTblSt, const WW8Fib& rFib)
-    : WW8PLCFx(rFib.GetFIBVersion(), false), pStatus(0), nIsEnd(0), nBookmarkId(1)
+    : WW8PLCFx(rFib.GetFIBVersion(), false), nIsEnd(0), nBookmarkId(1)
 {
     if( !rFib.fcPlcfbkf || !rFib.lcbPlcfbkf || !rFib.fcPlcfbkl ||
         !rFib.lcbPlcfbkl || !rFib.fcSttbfbkmk || !rFib.lcbSttbfbkmk )
@@ -3997,14 +3997,12 @@ WW8PLCFx_Book::WW8PLCFx_Book(SvStream* pTblSt, const WW8Fib& rFib)
             nIMax = pBook[0]->GetIMax();
         if( pBook[1]->GetIMax() < nIMax )
             nIMax = pBook[1]->GetIMax();
-        pStatus = new eBookStatus[ nIMax ];
-        memset( pStatus, 0, nIMax * sizeof( eBookStatus ) );
+        aStatus.resize(nIMax);
     }
 }
 
 WW8PLCFx_Book::~WW8PLCFx_Book()
 {
-    delete[] pStatus;
     delete pBook[1];
     delete pBook[0];
 }
@@ -4118,16 +4116,18 @@ long WW8PLCFx_Book::GetLen() const
 
 void WW8PLCFx_Book::SetStatus(sal_uInt16 nIndex, eBookStatus eStat )
 {
-    OSL_ENSURE(nIndex < nIMax, "set status of non existing bookmark!");
-    pStatus[nIndex] = (eBookStatus)( pStatus[nIndex] | eStat );
+    SAL_WARN_IF(nIndex >= nIMax, "sw.ww8",
+                "bookmark index " << nIndex << " invalid");
+    eBookStatus eStatus = aStatus.at(nIndex);
+    aStatus[nIndex] = static_cast<eBookStatus>(eStatus | eStat);
 }
 
 eBookStatus WW8PLCFx_Book::GetStatus() const
 {
-    if( !pStatus )
+    if (aStatus.empty())
         return BOOK_NORMAL;
     long nEndIdx = GetHandle();
-    return ( nEndIdx < nIMax ) ? pStatus[nEndIdx] : BOOK_NORMAL;
+    return ( nEndIdx < nIMax ) ? aStatus[nEndIdx] : BOOK_NORMAL;
 }
 
 long WW8PLCFx_Book::GetHandle() const
diff --git a/sw/source/filter/ww8/ww8scan.hxx b/sw/source/filter/ww8/ww8scan.hxx
index a323b84..e9df6cb2 100644
--- a/sw/source/filter/ww8/ww8scan.hxx
+++ b/sw/source/filter/ww8/ww8scan.hxx
@@ -703,8 +703,8 @@ class WW8PLCFx_Book : public WW8PLCFx
 {
 private:
     WW8PLCFspecial* pBook[2];           // Start and End Position
-    ::std::vector<String> aBookNames;   // Name
-    eBookStatus* pStatus;
+    std::vector<String> aBookNames;   // Name
+    std::vector<eBookStatus> aStatus;
     long nIMax;                         // Number of Booknotes
     sal_uInt16 nIsEnd;
     int nBookmarkId; // counter incremented by GetUniqueBookmarkName.
-- 
1.9.1

